packages/lodestar/src/network/gossip/gossip.ts

```diff
@@ -0,0 +1,199 @@
+import {INetworkOptions} from "../options";
+import {GossipEventEmitter, IGossip, IGossipEvents, IGossipModules,} from "./interface";
+import {GossipEvent} from "./constants";
+import {handleIncomingBlock} from "./handlers/block";
+import {handleIncomingAttestation} from "./handlers/attestation";
+import {handleIncomingAttesterSlashing} from "./handlers/attesterSlashing";
+import {handleIncomingProposerSlashing} from "./handlers/proposerSlashing";
+import {handleIncomingVoluntaryExit} from "./handlers/voluntaryExit";
+import {handleIncomingAggregateAndProof} from "./handlers/aggregateAndProof";
+import {toHex} from "@chainsafe/eth2.0-utils";
+export class Gossip extends (EventEmitter as { new(): GossipEventEmitter }) implements IGossip {
+
+  protected readonly  opts: INetworkOptions;
+  protected readonly config: IBeaconConfig;
+  protected readonly  libp2p: LibP2p;
+  protected readonly  pubsub: Gossipsub;
+  protected readonly  logger: ILogger;
+
+  private handlers: Map<string, Function>;
+
+  public constructor(opts: INetworkOptions, {config, libp2p, logger}: IGossipModules) {
+    super();
+    this.opts = opts;
+    this.config = config;
+    this.libp2p = libp2p;
+    this.logger = logger.child({module: "gossip", level: LogLevel[logger.level]});
+    this.logger.silent = logger.silent;
+    this.pubsub = new Gossipsub(libp2p, {gossipIncoming: false});
+    this.handlers = this.registerHandlers();
+  }
+
+  public async start(): Promise<void> {
+    await promisify(this.pubsub.start.bind(this.pubsub))();
+    this.handlers.forEach((handler, topic) => {
+      this.pubsub.on(topic, handler);
+    });
+  }
+
+  public async stop(): Promise<void> {
+    await promisify(this.pubsub.stop.bind(this.pubsub))();
+    this.handlers.forEach((handler, topic) => {
+      this.pubsub.removeListener(topic, handler);
+    });
+  }
+
+  public async publishBlock(block: BeaconBlock): Promise<void> {
+    await promisify(this.pubsub.publish.bind(this.pubsub))(
+      getGossipTopic(GossipEvent.BLOCK), serialize(block, this.config.types.BeaconBlock)
+    );
+    this.logger.verbose(`Publishing block at slot: ${block.slot}`);
+  }
+
+  public async publishCommiteeAttestation(attestation: Attestation): Promise<void> {
+    const subnet = getAttestationSubnet(attestation);
+    await promisify(this.pubsub.publish.bind(this.pubsub))(
+      getAttestationSubnetTopic(attestation), serialize(attestation, this.config.types.Attestation));
+    //backward compatible
+    await promisify(this.pubsub.publish.bind(this.pubsub))(
+      getGossipTopic(GossipEvent.ATTESTATION), serialize(attestation, this.config.types.Attestation)
+    );
+    this.logger.verbose(
+      `Publishing attestation ${toHex(hashTreeRoot(attestation, this.config.types.Attestation))} for subnet ${subnet}`
+    );
+  }
+
+  public async publishAggregatedAttestation(aggregateAndProof: AggregateAndProof): Promise<void> {
```

Comment:
What I read from the latest validator spec https://github.com/ethereum/eth2.0-specs/blob/v09x/specs/validator/0_beacon-chain-validator.md#aggregateandproof, validator gonna broadcast aggregated attestation. If that's the case then we may need to make this network gossip as a separate package?

Please correct me if I'm wrong.

Reply:
you got it wrong, validator publishes attestation using rest api, which means beacon node will publish it on the right topic. Validator is not connected to libp2p